### ISSUE TITLE
Memory usage information for each listed running notebook, in "Running" tab

### ISSUE THREAD

**Issue by ianshmean** [jupyter/notebook#3117]
It would be helpful to have memory usage information for each listed running notebook, in "Running" tab to help manage on memory usage.
I find myself having to keep System Monitor open to keep a check on ram usage.

Also, memory usage beside the kernel info at the top of an open notebook could be helpful

**Comment by takluyver** [comment]
I think [psutil](https://pypi.python.org/pypi/psutil/) can give us that information, but it's not always easy to install, so it might be best as an optional dependency, or an extension.

**Comment by blink1073** [comment]
It has [wheels](https://pypi.python.org/pypi/psutil) for windows and is available on [conda](https://anaconda.org/search?q=psutil) defaults and conda-forge.

**Comment by takluyver** [comment]
Right, but no wheels for Mac or Linux.

**Comment by blink1073** [comment]
Ah, right.

**Comment by blink1073** [comment]
Works fine on macOS:

```bash
$ pip install psutil
Collecting psutil
  Downloading psutil-5.4.2.tar.gz (411kB)
    100% |████████████████████████████████| 419kB 569kB/s
Building wheels for collected packages: psutil
  Running setup.py bdist_wheel for psutil ... done
  Stored in directory: /Users/ssilvester/Library/Caches/pip/wheels/62/36/0b/55b377328e81ae7345cd9a4b7b9d90c423c6cfad7d9c9b08b0
Successfully built psutil
Installing collected packages: psutil
Successfully installed psutil-5.4.2
```

**Comment by blink1073** [comment]
Ditto on Ubuntu 16.04.

**Comment by takluyver** [comment]
Works fine for you, but you've presumably got a compiler and the Python header files installed already. ;-)

Also, even if there are wheels for the three major platforms, that doesn't cover installing on e.g. a raspberry pi. The excellent [piwheels](https://www.piwheels.hostedpi.com/) project might help for that particular platform, but there are still others out there which don't have something like that.

Compiled dependencies have got much more tractable in recent years, but they're still a pain often enough that I'd rather avoid making them compulsory if possible.

**Comment by blink1073** [comment]
Fair enough

**Comment by blink1073** [comment]
Ye olde "works on my (virtual) machine"  :smile:

**Comment by takluyver** [comment]
To be fair, it may work for 95% of users. But when the failure mode is "can't install your software", even a small percentage matters.

**Comment by ianshmean** [comment]
So it seems like the consensus is that this would be more effort than it's worth. 
Happy to close if that's right?

**Comment by takluyver** [comment]
It's the kind of thing I think would be best developed as an extension first, though I don't know how easy it would be to extend the notebook in that way.

**Comment by yshkk** [comment]
I tried to write a custom handler to solve this problem : 
```python
import subprocess
import re
from tornado import gen, web
from notebook.base.handlers import APIHandler
import psutil

KERNEL_REGEX = re.compile(r'.+kernel-(.+)\.json')
PID_REGEX = re.compile(r'(\d+).+')


class KernelMemoryHandler(APIHandler):
    '''
    get memory of each kernel
    '''
    @web.authenticated
    @gen.coroutine

    def get(self):
        ps = subprocess.Popen(('ps', 'ax'), stdout=subprocess.PIPE)
        output = ps.communicate()[0]
        result={}
        for line in output.split('\n'):
            if 'kernel-' in line:
                kernel_id = re.sub(KERNEL_REGEX, r'\1', line)
                pid = int(re.sub(PID_REGEX, r'\1', line))
                process = psutil.Process(pid)
                mem = int(process.memory_info()[0])
                result[kernel_id]={'pid':pid,'memory': mem}
        self.finish(result)
       
default_handlers = [
    (r"/api/kernel/memory", KernelMemoryHandler)
]
```
it works on my Mac and Linux server. 😄

**Comment by psychemedia** [comment]
Searching around, there's also this example from several years ago of a notebook kernel memory reporter: https://github.com/iiSeymour/nbtop

**Comment by ggrrll** [comment]
maybe this SO post could give some more ideas ...

https://stackoverflow.com/questions/34685825/jupyter-notebook-memory-usage-for-each-notebook

**Comment by stefansimik** [comment]
It would be really useful to see memory consumption for each notebook.
We have the same problem - some notebooks are no more used and can consume large chunk of memory. There is no easy way how to identify these notebooks, to close them and shutdown their kernels.

We use JupyterLab in shared environment and it is a real daily pain.
If we could see, how much memory is consumed by each notebook, it would make our life much easier every day.

**Comment by julianofischer** [comment]
What about https://github.com/yuvipanda/nbresuse ?
It is working well here.

**Comment by stefansimik** [comment]
@julianofischer Thank you,
nbresuse looks great, 
but it looks like it is not available in JupyterLab,
only for the jupyter-notebook interface... Is it true?

**Comment by psychemedia** [comment]
@julianofischer I'm pretty sure I've had a version of it running in a n old notebook server for some time...

**Comment by igorng** [comment]
> @julianofischer Thank you,
> nbresuse looks great,
> but it looks like it is not available in JupyterLab,
> only for the jupyter-notebook interface... Is it true?

Yeah I think it is not available on jupyterLab. I root for something like this too. And exposed like prometeus metrics would be great.

**Comment by greatvovan** [comment]
> So it seems like the consensus is that this would be more effort than it's worth.

Disagree. Not everybody runs notebooks locally. You may need to run a notebook on a remote server, in Docker, etc. These options may be harder to examine with standard OS tools, moreover all running kernels look like just standard Python processes, it is impossible to understand which process runs which kernel.

Running additional code to find out memory usage is very inconvenient. Furthermore, a kernel may be currently busy with executing a cell's code. Ability to monitor memory consumption "outside" of the kernel would be very appreciated.

**Comment by kevin-bates** [comment]
I agree that this would be a useful feature and that it presents some challenges that might warrant this issue being moved to jupyter_server given the current maintenance level of this repo.  

As @greatvovan points out, we cannot always assume the kernel process is local, so I believe the location for where the memory usage is gathered must be at the kernel manager level which then implies this be something implemented in jupyter_client and surfaced to notebook users via the REST API. 

((Btw, this is a perfect example of how [kernel providers](https://github.com/jupyter/enhancement-proposals/pull/45) could be utilized since each provider would be the entity that knows how to gather memory usage for its set of kernels - kubernetes, docker, remote YARN, local, etc. - and we could add a stats "capture" method to the interface that providers optionally implement.))

Given this issue is not going to be addressed in this repo, are others okay with its transfer to jupyter_server?  (I'll hold off the transfer for 48 hours or so to allow for discussion.)
